gcp/config.go

```diff
@@ -0,0 +1,240 @@
+
+// Items
+type item struct {
+	Key   string `yaml:"key"`
+	Value string `yaml:"value"`
+}
+
+// tags
+type tags struct {
+	Items []string `yaml:"items"`
+}
+
+// disk
+type disk struct {
+	DeviceName       string           `yaml:"deviceName"`
+	Type             string           `yaml:"type"`
+	Boot             bool             `yaml:"boot"`
+	AutoDelete       bool             `yaml:"autoDelete"`
+	InitializeParams initializeParams `yaml:"initializeParams"`
+}
+
+// networkInterface
+type networkInterface struct {
+	Network       string         `yaml:"network"`
+	AccessConfigs []accessConfig `yaml:"accessConfigs"`
+}
+
+// accessConfig
+type accessConfig struct {
+	Name  string `yaml:"name"`
+	NatIP string `yaml:"natIP"`
+}
+
+// Config
+type Config struct {
+	projectID string      `yaml:"-"`
+	Resources []resources `yaml:"resources"`
+	Outputs   []output    `yaml:"outputs"`
+}
+
+// resources
+type resources struct {
+	Type       string     `yaml:"type"`
+	Name       string     `yaml:"name"`
+	Properties properties `yaml:"properties"`
+}
+
+// properties
+type properties struct {
+	Zone              string             `yaml:"zone"`
+	MachineType       string             `yaml:"machineType"`
+	Metadata          metadata           `yaml:"metadata"`
+	Tags              tags               `yaml:"tags"`
+	Disks             []disk             `yaml:"disks"`
+	NetworkInterfaces []networkInterface `yaml:"networkInterfaces"`
+}
+
+// initializeParams
+type initializeParams struct {
+	SourceImage string `yaml:"sourceImage"`
+}
+
+// output
+type output struct {
+	Name  string `yaml:"name"`
+	Value string `yaml:"value"`
+}
+
+func NewConfig(projectID string) Config {
+	c := Config{
+		projectID: projectID,
+		Resources: []resources{
+			{
+				Type: "compute.v1.instance",
+				Name: "calyptia-core-instance",
+				Properties: properties{
+					Zone:        DefaultZone,
+					MachineType: fmt.Sprintf("projects/%s/zones/%s/machineTypes/%s", projectID, DefaultZone, DefaultMachineType),
+					Metadata:    metadata{Items: []item{}},
+					Tags:        tags{Items: []string{"calyptia-core-instance"}},
+					Disks: []disk{
+						{
+							DeviceName: "boot",
+							Type:       "PERSISTENT",
+							Boot:       true,
+							AutoDelete: true,
+							InitializeParams: initializeParams{
+								SourceImage: BaseCalyptiaImage,
+							},
+						},
+					},
+					NetworkInterfaces: []networkInterface{
+						{
+							Network: fmt.Sprintf("https://www.googleapis.com/compute/v1/projects/%s/global/networks/default", projectID),
+							AccessConfigs: []accessConfig{
+								{
+									Name: "External NAT",
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+		Outputs: []output{
+			{
+				Name:  "IPAddress",
+				Value: "$(ref.calyptia-core-instance.networkInterfaces[0].accessConfigs[0].natIP)",
+			},
+		},
+	}
+	return c
+}
+
+func (c *Config) SetZone(zone string) *Config {
+	if zone == "" {
+		return nil
+	}
+	c.Resources[0].Properties.Zone = zone
```

Comment:
This could panic. Missing a nigl check.

Reply:
im not checking this index bcz is the one that we create by default, so dont will have less than [0], you think we still need to check?

Reply:
Resources is an exported field so it can be set to anything. Can perfectly be nil or an empty slice.